### PR TITLE
Implement new parameter `upgrade_segments` for `OPTIMIZE` statement

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/OptimizeTableAnalyzer.java [hunk 1]
```diff
@@ -90,4 +92,22 @@ public OptimizeTableAnalyzedStatement analyze(OptimizeStatement stmt, Analysis a
         }
         return indexNames;
     }
+
+    private void validateSettings(Settings settings) {
+        if (settings.getAsBoolean(UPGRADE_SEGMENTS.name(), UPGRADE_SEGMENTS.defaultValue())) {
+            if (!settings.getAsBoolean(FLUSH.name(), FLUSH.defaultValue())) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/OptimizeTableAnalyzer.java]: is it really needed/worth doing the concrete checks? wouldn't it be enough to check that `upgrade_segments` is the only setting (by checking the settings size additionally)?

File: sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java [hunk 2]
```diff
@@ -103,22 +105,32 @@ public DDLStatementDispatcher(Provider<BlobAdminClient> blobAdminClient,
...
+                request.flush(analysis.settings().getAsBoolean(OptimizeSettings.FLUSH.name(),
+                    ForceMergeRequest.Defaults.FLUSH));
+
+                request.indicesOptions(IndicesOptions.lenientExpandOpen());
+
+                FutureActionListener<ForceMergeResponse, Long> listener =
+                    new FutureActionListener<>(Functions.constant((long) analysis.indexNames().size()));
+                transportActionProvider.transportForceMergeAction().execute(request, listener);
+                return listener;
+            }
```
Comment (seut) [inline sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java]: maybe it makes sense to move both to concrete methods (executeMergeSegments & executeUpgradeSegments)

File: sql/src/test/java/io/crate/integrationtests/OptimizeTableIntegrationTest.java [hunk 3]
```diff
@@ -101,6 +101,24 @@ public void testOptimizeWithParams() throws Exception {
...
+        assertThat((long) response.rows()[0][0], lessThanOrEqualTo(3L));
+
+        execute("delete from test where id=1");
+        refresh();
+        execute("optimize table test with (max_num_segments=1, only_expunge_deletes=true)");
+        assertThat(response.rowCount(), is(1L));
+        assertThat(response.rows().length, is(0));
+
+        execute("select count(*) from test");
+        assertThat(response.rows()[0][0], is(2L));
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/OptimizeTableIntegrationTest.java]: this test seems not to test the `upgrade_segments` parameter as the test name implies to

File: sql/src/test/java/io/crate/integrationtests/SysShardMinLuceneVersionTest.java [hunk 4]
```diff
@@ -76,4 +71,36 @@ public void testMinLuceneVersion() throws Exception {
...
                "test_upgrade_required_parted| UNASSIGNED| NULL| 10\n"));
     }
+
+    @Test
+    public void testUpgradeSegments() throws Exception {
+        startUpNodeWithDataDir("/indices/cluster_checks/cratedata_lucene_min_version.zip");
+        execute("select table_name, routing_state, min_lucene_version, count(*) from sys.shards " +
+                "where table_name IN " +
+                "('test_upgrade_required', 'test_upgrade_required_parted', 'test_blob_upgrade_required') " +
+                "group by table_name, routing_state, min_lucene_version order by 1, 2, 3");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/SysShardMinLuceneVersionTest.java]: test would make more sense if query filters just old segments and upgrades them (no old segments after upgrading at all)

File: sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java [hunk 5]
```diff
@@ -67,13 +72,17 @@ public DDLStatementDispatcher(Provider<BlobAdminClient> blobAdminClient,
                                   AlterTableOperation alterTableOperation,
                                   RepositoryService repositoryService,
                                   SnapshotRestoreDDLDispatcher snapshotRestoreDDLDispatcher,
-                                  TransportActionProvider transportActionProvider) {
+                                  Provider<TransportUpgradeAction> transportUpgradeActionProvider,
```
Comment (seut) [inline sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java]: do we really need a provider here?

File: sql/src/test/java/io/crate/integrationtests/OptimizeTableIntegrationTest.java [hunk 6]
```diff
@@ -101,6 +101,24 @@ public void testOptimizeWithParams() throws Exception {
...
+    public void testOptimizeWithUpgradeSegments() throws Exception {
+        execute("create table test (id int primary key, name string)");
+        ensureYellow();
+        execute("insert into test (id, name) values (0, 'Trillian'), (1, 'Ford'), (2, 'Zaphod')");
+        execute("select count(*) from test");
+        assertThat((long) response.rows()[0][0], lessThanOrEqualTo(3L));
+
+        execute("delete from test where id=1");
+        refresh();
+        execute("optimize table test with (upgrade_segments=true)");
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/OptimizeTableIntegrationTest.java]: I'm not sure if this test makes sense at all. the `testUpgradeSegments` test is already doing a integration test, either move is to here, or just leave and delete this test.

File: sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java (hunk 5 above)
Comment (matriv) [inline sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java]: Yep otherwise we get guice cyclic dependency errors

**Pull request merged**